[On-week] Add persisting page size and sorting for Management tables

ElenaStoeva opened the pull request:
Addresses https://github.com/elastic/kibana/issues/56406

## Summary

This PR is part of my June 2024 On-week project. It adds functionality for persisting table page size (rows per page) and sorting in EUI tables. When a user changes the size or sort, the new values are saved in local storage, so that when the user navigates out of the page and comes back, the page size and sort will be the same.

In this PR, the functionality is added to the following tables:
- Ingest Pipelines
- Index Management - all tables (indices, data streams, index templates, component templates, enrich policies)
- ILM policies



https://github.com/elastic/kibana/assets/59341489/08b0c004-65a1-4a58-b8fb-99010e1c3248



<!--
### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
-->

ElenaStoeva committed:
[Spacetime] Add persisting page size for Management tables

ElenaStoeva committed:
Add persisting page size to IM and ILM

ElenaStoeva committed:
Use a different page size local storage for each table

ElenaStoeva committed:
Move onTableChange handler to hook

ElenaStoeva committed:
Add persistent sort

ElenaStoeva committed:
Fix local storage variable delimeter

ElenaStoeva committed:
Add initial sort for enrich policies table

ElenaStoeva committed:
Merge branch 'main' into on-week/persisting-table-page-size

ElenaStoeva committed:
Refactor changes, add validation and tests

ElenaStoeva commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

ElenaStoeva committed:
Add more documentation and some fixes

ElenaStoeva commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

ElenaStoeva commented:
/ci

ElenaStoeva committed:
Fix failing tests and return original page size options in tables

ElenaStoeva committed:
Merge branch 'on-week/persisting-table-page-size' of https://github.com/ElenaStoeva/kibana into on-week/persisting-table-page-size

ElenaStoeva commented:
/ci

ElenaStoeva committed:
Fix ILM test failures

ElenaStoeva committed:
Merge branch 'main' into on-week/persisting-table-page-size

ElenaStoeva commented:
/ci

ElenaStoeva committed:
Fix type error

ElenaStoeva commented:
/ci

ElenaStoeva committed:
Small fixes

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

elasticmachine commented:
Pinging @elastic/appex-sharedux (Team:SharedUX)

ElenaStoeva committed:
Merge branch 'main' into on-week/persisting-table-page-size

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

ElenaStoeva committed:
Fix type errors

ElenaStoeva committed:
Resolve merge conflicts

ElenaStoeva committed:
Update readme

yuliacech commented on packages/shared-ux/table_persist/src/use_table_persist.ts:
```diff
@@ -0,0 +1,81 @@
+ * Returns the persisting page size and sort and the onTableChange handler that should be passed
+ * as props to an Eui table component.
+ */
+export const useEuiTablePersist = <T extends object>({
```
I don't think this function should be called a hook and start with `use` because it doesn't contain any react hooks. It looks like a regular function that returns some data and a function.

ElenaStoeva replied:
Indeed it wasn't a hook! After the latest changes ([096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b)), it uses React built-in hooks so I think now we can call it a hook 😄

yuliacech commented on packages/shared-ux/table_persist/src/use_table_persist.ts:
```diff
@@ -0,0 +1,81 @@
+  pageSizeOptions,
+}: EuiTablePersistProps<T>) => {
+  const storage = createStorage({
+    engine: window.localStorage,
```
I think this level of abstraction from localStorage is needed if the consumer of this package can provide a storage engine themselves. Otherwise if the storage engine is hard coded, I think the engine might as well be hidden in the storage class.

sebelga replied:
I think we could use a `useCallback` to wrap `onTableChange` that is returned to the consumer.

We could also have `sorting` and `pageSize` that are returned to the consumer as 2 states (their values would be updated in `onTableChange`) so they can react to their value change if need be?

ElenaStoeva replied:
Good calls! Removed the engine abstraction from the Storage class and added `useCallback` and states with [096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b).

yuliacech reviewed (approved):
That is a great improvement to the tables, @ElenaStoeva! 
I verified locally that the page size is persisted between page loads. Code changes LGTM, I would just recommend not naming the function a hook if it doesn't use any react hooks. Also I think some unit tests for the `useEuiTablePersist` could be useful as well.

sebelga commented on x-pack/plugins/index_lifecycle_management/public/application/sections/policy_list/components/policy_table.tsx:
```diff
@@ -103,6 +104,17 @@ export const PolicyTable: React.FunctionComponent<Props> = ({ policies }) => {
   );
   const { setListAction } = usePolicyListContext();
 
+  const PAGE_SIZE_OPTIONS = [10, 25, 50];
```
It is probably best to keep this constant outside of the component, next to `SHOW_MANAGED_POLICIES_BY_DEFAULT` for example.

ElenaStoeva replied:
Addressed in [096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b).

sebelga commented on x-pack/plugins/index_management/public/application/components/component_templates/component_template_list/table.tsx:
```diff
@@ -88,6 +89,14 @@ export const ComponentTable: FunctionComponent<Props> = ({
 
   const [selection, setSelection] = useState<ComponentTemplateListItem[]>([]);
 
+  const PAGE_SIZE_OPTIONS = [10, 20, 50];
```
Same comment here to keep the constant outside of the component.

sebelga commented on packages/shared-ux/table_persist/src/use_table_persist.ts:
```diff
@@ -0,0 +1,81 @@
+ */
+
+import { Criteria } from '@elastic/eui';
+import { PropertySort } from '@elastic/eui/src/services/sort/property_sort';
```
I would rather not deep import in case the path changes.

We can create our own type for it

```ts
type PropertySort = Criteria<any>['sort'];
```

ElenaStoeva replied:
That makes sense! I added this type in `types.ts` ([096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b))

sebelga commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_table/data_stream_table.tsx:
```diff
@@ -282,6 +271,22 @@ export const DataStreamTable: React.FunctionComponent<Props> = ({
     ],
   };
 
+  const PAGE_SIZE_OPTIONS = [10, 20, 50];
```
Same here for the constant

sebelga commented on x-pack/plugins/index_management/public/application/sections/home/template_list/template_table/template_table.tsx:
```diff
@@ -188,17 +189,21 @@ export const TemplateTable: React.FunctionComponent<Props> = ({
   ];
 
-  const pagination = {
+  const PAGE_SIZE_OPTIONS = [10, 20, 50];
```
Same here for the constant

sebelga commented on x-pack/plugins/ingest_pipelines/public/application/sections/pipelines_list/table.tsx:
```diff
@@ -115,6 +116,14 @@ export const PipelineTable: FunctionComponent<Props> = ({
   const { history } = useKibana().services;
   const [selection, setSelection] = useState<Pipeline[]>([]);
 
+  const PAGE_SIZE_OPTIONS = [10, 20, 50];
```
Same here for the constant

loixlab reviewed (commented):
Great work @ElenaStoeva ! I left a few comments regarding code style and readability.
As @yuliacech mentioned currently it is not a hook and I think we could return at least the `onTableChange` in a `useCallback` and probably make the 2 other value a state so those are reactive values for consumers.

ElenaStoeva committed:
Address feedback

ElenaStoeva commented on packages/shared-ux/table_persist/src/storage.ts:
```diff
@@ -0,0 +1,48 @@
+
+type IStorageEngine = typeof window.localStorage;
+
+export class Storage {
```
Nice catch! Fixed in [096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b).

ElenaStoeva commented on packages/shared-ux/table_persist/src/use_table_persist.ts:
```diff
@@ -0,0 +1,81 @@
+  /** (Optional) Specifies a custom initial table sorting. */
+  initialSort?: PropertySort;
+  /** (Optional) Specifies a custom initial page size for the table.
+   * Defaults to {@link DEFAULT_INITIAL_PAGE_SIZE} */
```
Yeah, this makes sense! Addressed in [096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b).

ElenaStoeva commented on packages/shared-ux/table_persist/src/validate_persist_data.ts:
```diff
@@ -0,0 +1,38 @@
+  if (sort) {
+    const field = sort.field;
+    const direction = sort.direction;
+    if (!(typeof field === 'string') || !(direction === 'asc' || direction === 'desc')) {
```
Addressed in [096abab](https://github.com/elastic/kibana/pull/186997/commits/096ababced9dd39a51756bebea3b67b1dc95650b).

ElenaStoeva commented:
Many thanks for the thorough reviews, @sebelga and @yuliacech!
I addressed your comments and added some unit tests for the (now already) hook. 😅 Please let me know if I missed anything or if you have further suggestions.

ElenaStoeva commented on packages/shared-ux/table_persist/src/use_table_persist.ts:
```diff
@@ -0,0 +1,81 @@
+    const newPersistData: PersistData<T> = {};
+    newPersistData.pageSize = nextValues.page?.size ?? storagePageSize;
+    const newSort = nextValues?.sort;
+    newPersistData.sort = newSort?.field
```
Sure, I updated it as suggested, with some modifications due to some recent changes.

On a side note, I just want to note that with the current logic, if the consumer provides an initial sort, when the user removes a sort (the change criteria would contain just the `direction` property, with `undefined` field), then we store `undefined` sort in local storage, so next time when the user visits this page, the initial sort will be restored (rather than no sort to be preserved). Just wanted to make sure this is the desired behaviour.

ElenaStoeva committed:
Fix type errors

ElenaStoeva committed:
Merge branch 'main' into on-week/persisting-table-page-size

sebelga commented on packages/shared-ux/table_persist/src/use_table_persist.ts:
```diff
@@ -0,0 +1,87 @@
+        storage.set(tableId, nextPersistData);
+      }
+    },
+    [customOnTableChange, pageSize, sort, storage, storagePageSize, storageSort, tableId]
```
In a callback that it setting the state (`setPageSize`, `setSort`), it is usually better to avoid having the state being set as a dependency. It could create an infinite loop. 
Here I checked and there are no infinite loop but I think we could avoid it this way

```ts
const onTableChange = useCallback(
  (nextValues: Criteria<T>) => {
    if (customOnTableChange) {
      customOnTableChange(nextValues);
    }

    let nextSort: PropertySort<T> | undefined;
    if (nextValues.sort?.field && nextValues.sort?.direction) {
      // Both field and direction are needed for a valid sort criteria
      nextSort = nextValues.sort;
    }

    if (nextValues.sort?.field || nextValues.sort?.direction) {
      setSort(nextSort);
    }

    const nextPageSize = nextValues.page?.size;
    if (nextPageSize) {
      setPageSize(nextPageSize);
    }

    if (
      (nextPageSize && nextPageSize !== storagePageSize) ||
      (nextSort && nextSort !== storageSort)
    ) {
      const nextPersistData: PersistData<T> = {
        pageSize: nextPageSize,
        sort: nextSort,
      };
      storage.set(tableId, nextPersistData);
    }
  },
  [customOnTableChange, storage, storagePageSize, storageSort, tableId]
);
```

ElenaStoeva replied:
Good call! I added your suggestion with a small modification to ensure that sort would be cleaned up in local storage if the user removes the sort on a specific field (when this happens, `field` is an empty string). [7754207](https://github.com/elastic/kibana/pull/186997/commits/77542076080846eb27e852561a1e43f8f3b75af5)

loixlab reviewed (approved):
Looking good! I left one more comment to avoid having the state as a dependency of a callback setting the state.
I'll approve to unblock the PR but it'd be great to apply my suggestion 🙏  Thanks!

ElenaStoeva committed:
Address feedback

ElenaStoeva committed:
Fix page size url params in Index management

ElenaStoeva committed:
Merge branch 'main' into on-week/persisting-table-page-size
